src/Workspaces/Remote/ServiceHub/Services/SnapshotService.JsonRpcAssetSource.cs

```diff
@@ -34,10 +34,32 @@ public override async Task<IList<(Checksum, object)>> RequestAssetsAsync(int sco
             {
                 using (RoslynLogger.LogBlock(FunctionId.SnapshotService_RequestAssetAsync, GetRequestLogInfo, scopeId, checksums, cancellationToken))
                 {
-                    return await _owner.Rpc.InvokeAsync(WellKnownServiceHubServices.AssetService_RequestAssetAsync,
-                        new object[] { scopeId, checksums.ToArray() },
-                        (s, c) => ReadAssets(s, scopeId, checksums, c), cancellationToken).ConfigureAwait(false);
+                    try
+                    {
+                        return await _owner.RunServiceAsync(() =>
+                        {
+                            return _owner.Rpc.InvokeAsync(WellKnownServiceHubServices.AssetService_RequestAssetAsync,
+                                new object[] { scopeId, checksums.ToArray() },
+                                (s, c) => ReadAssets(s, scopeId, checksums, c), cancellationToken);
+                        }, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception ex) when (ReportUnlessCanceled(ex, cancellationToken))
+                    {
+                        throw ExceptionUtilities.Unreachable;
+                    }
+                }
+            }
+
+            private bool ReportUnlessCanceled(Exception ex, CancellationToken cancellationToken)
+            {
+                if (!cancellationToken.IsCancellationRequested &&
+                    ((IDisposableObservable)_owner.Rpc).IsDisposed)
+                {
+                    // kill OOP if snapshot service got disconnected due to this exception.
```

Comment:
we explicitly crash OOP process if back channel to move around solution data is closed. this should never closed in normal situation since we share 1 connection for all data synching.